src/ring_theory/finiteness.lean

```diff
@@ -182,8 +189,64 @@ begin
     exact finite_type.of_surjective (finite_type.mv_polynomial R (fin n)) f hsur }
 end
 
+/-- A finitely presented algebra is of finite type. -/
+lemma of_finitely_presented : finitely_presented R A → finite_type R A :=
+begin
+  rintro ⟨n, ⟨f, hf⟩⟩,
+  apply (finite_type.iff_quotient_mv_polynomial'').2,
+  use [n, f, hf.1]
+end
+
 end finite_type
 
+namespace finitely_presented
+
+theorem spec [h : finitely_presented R A] :
+  ∃ (n : ℕ) (f : mv_polynomial (fin n) R →ₐ[R] A),
+  surjective f ∧ (ring_hom.ker (alg_hom.to_ring_hom f)).fg :=
+h
+
+/-- If `e : A ≃ₐ[R] B` and `A` is finitely presented, then so is `B`. -/
+lemma equiv [finitely_presented R A] (e : A ≃ₐ[R] B) : finitely_presented R B :=
+begin
+  obtain ⟨n, f, hf⟩ := spec R A,
+  use [n, alg_hom.comp ↑e f],
+  split,
+  { exact function.surjective.comp e.surjective hf.1 },
+  suffices hker : (alg_hom.comp ↑e f).to_ring_hom.ker = f.to_ring_hom.ker,
+  { rw hker, exact hf.2 },
+  { have hco : (alg_hom.comp ↑e f).to_ring_hom = ring_hom.comp ↑e.to_ring_equiv f.to_ring_hom,
+    { have h : (alg_hom.comp ↑e f).to_ring_hom = e.to_alg_hom.to_ring_hom.comp f.to_ring_hom := rfl,
+      have h1 : ↑(e.to_ring_equiv) = (e.to_alg_hom).to_ring_hom := rfl,
+      rw [h, h1] },
+    rw [ring_hom.ker_eq_comap_bot, hco, ← ideal.comap_comap, ← ring_hom.ker_eq_comap_bot,
+      ring_hom.ker_coe_equiv (alg_equiv.to_ring_equiv e), ring_hom.ker_eq_comap_bot] }
+end
+
+/-- The ring of polynomials in finitely many variables is finitely presented. -/
+lemma mv_polynomial (ι : Type u_2) [fintype ι] : finitely_presented R (mv_polynomial ι R) :=
+begin
+  obtain ⟨n, equiv⟩ := @fintype.exists_equiv_fin ι _,
+  replace equiv := mv_polynomial.alg_equiv_of_equiv R (nonempty.some equiv),
+  use [n, alg_equiv.to_alg_hom equiv.symm],
+  split,
+  { exact (alg_equiv.symm equiv).surjective },
+  suffices hinj : function.injective (alg_hom.to_ring_hom (alg_equiv.to_alg_hom equiv.symm)),
```

Comment:
More projections:
```suggestion
  suffices hinj : function.injective (equiv.symm.to_alg_hom.to_ring_hom),
```

Reply:
(You can remove the rest of the parentheses then too.)